### PR TITLE
Add also_check_avg

### PR DESCRIPTION
**Pull request by motbob**
Sometimes, video changes come in the form of overall brightness changes, like fog in censored scenes or anti-seizure filters, which Max/Min isn't good at catching. This change attempts to catch those scenes. Tested to have no false positives in a comparison between Abema and Crunchyroll.

Speed hit is something like 5%, maybe?

Update from previous pull request: adjust the avg diff threshold based on the max/min threshold. If the max/min thr is less conservative, the avg diff threshold can also be less conservative. Values were derived by finding the point at which the first false positive appeared for a given thr and then adding a 50% buffer. Also, fix bug that prevented users from using avg threshold (`thr` less than 1) at all.

### CODE REVIEW

File: lvsfunc/comparison.py [hunk 1]
```diff
@@ -549,6 +549,8 @@ def diff(*clips: vs.VideoNode,
          height: int = 288,
          interleave: bool = False,
          return_ranges: bool = False,
+         also_check_avg: bool = False,
```
Comment (Setsugennoao) [inline lvsfunc/comparison.py:552]: Can you find better names?

File: lvsfunc/comparison.py [hunk 2]
```diff
@@ -605,8 +609,13 @@ def diff(*clips: vs.VideoNode,
     if (clips and len(clips) != 2) or (namedclips and len(namedclips) != 2):
         raise CustomValueError("Must pass exactly 2 `clips` or `namedclips`!", diff)
 
-    if not 1 <= thr < 128:
-        raise CustomValueError(f"`thr` must be between 1 and 128, not {thr}!", diff)
+    if not thr < 128:
+        raise CustomValueError(f"`thr` must be below 128!", diff)
```
Comment (Setsugennoao) [inline lvsfunc/comparison.py:613]: The error is missing the reason

File: lvsfunc/comparison.py [hunk 3]
```diff
@@ -605,8 +609,13 @@ def diff(*clips: vs.VideoNode,
     if (clips and len(clips) != 2) or (namedclips and len(namedclips) != 2):
         raise CustomValueError("Must pass exactly 2 `clips` or `namedclips`!", diff)
 
-    if not 1 <= thr < 128:
-        raise CustomValueError(f"`thr` must be between 1 and 128, not {thr}!", diff)
+    if not thr < 128:
```
Comment (Setsugennoao) [inline lvsfunc/comparison.py:612]: Why did you remove the positive number check?

File: lvsfunc/comparison.py [hunk 4]
```diff
@@ -605,8 +609,13 @@ def diff(*clips: vs.VideoNode,
...
+        raise CustomValueError(f"`thr` must be below 128!", diff)
+    
+    if also_check_avg_thr == None and thr > 1:
+        also_check_avg_thr = (128 - thr) * 0.000046875 + 0.0105
```
Comment (Setsugennoao) [inline lvsfunc/comparison.py:616]: Magic numbers?

File: lvsfunc/comparison.py [hunk 5]
```diff
@@ -645,10 +654,20 @@ def _cb(n: int, f: vs.VideoFrame) -> None:
...
-                frames.append(n)
+        if also_check_avg:
+            ps = core.std.PlaneStats(a, b)
+            def transfer_property(n, f):
+                fout = f[1].copy()
+                fout.props['PlaneStatsDiff'] = f[0].props['PlaneStatsDiff']
+                return fout
+            diff_clip = core.std.ModifyFrame(clip=ps, clips=[ps, diff_clip], selector=transfer_property)
+            def _cb(n: int, f: vs.VideoFrame) -> None:
+                if get_prop(f, 'PlaneStatsMin', typ) <= thr or get_prop(f, 'PlaneStatsMax', typ) >= 255 - thr > thr or get_prop(f, 'PlaneStatsDiff', float) > also_check_avg_thr:
```
Comment (Setsugennoao) [inline lvsfunc/comparison.py:665]: Duplicate code, extract it to another function

File: lvsfunc/comparison.py [hunk 6]
```diff
@@ -645,10 +654,20 @@ def _cb(n: int, f: vs.VideoFrame) -> None:
         assert diff_clip.format
 
         typ = float if diff_clip.format.sample_type == vs.FLOAT else int
-
-        def _cb(n: int, f: vs.VideoFrame) -> None:
-            if get_prop(f, 'PlaneStatsMin', typ) <= thr or get_prop(f, 'PlaneStatsMax', typ) >= 255 - thr > thr:
-                frames.append(n)
+        if also_check_avg:
+            ps = core.std.PlaneStats(a, b)
+            def transfer_property(n, f):
```
Comment (Setsugennoao) [inline lvsfunc/comparison.py:659]: Highly inefficient. Either use Expr or PropExpr

File: lvsfunc/comparison.py [hunk 7]
```diff
@@ -645,10 +654,20 @@ def _cb(n: int, f: vs.VideoFrame) -> None:
...
+                fout = f[1].copy()
+                fout.props['PlaneStatsDiff'] = f[0].props['PlaneStatsDiff']
+                return fout
+            diff_clip = core.std.ModifyFrame(clip=ps, clips=[ps, diff_clip], selector=transfer_property)
```
Comment (Setsugennoao) [inline lvsfunc/comparison.py:663]: Chained ModifyFrame and FrameEval is also inefficient, it can be merged.

File: lvsfunc/comparison.py [hunk 8]
```diff
@@ -549,6 +549,8 @@ def diff(*clips: vs.VideoNode,
          height: int = 288,
          interleave: bool = False,
          return_ranges: bool = False,
+         also_check_avg: bool = False,
+         also_check_avg_thr: float | None = None,
```
Comment (Setsugennoao) [inline lvsfunc/comparison.py:553]: These two params should be merged.

File: lvsfunc/comparison.py (hunk 2 above)
Comment (motbob) [reply lvsfunc/comparison.py:613]: sorry, I don't understand

File: lvsfunc/comparison.py (hunk 3 above)
Comment (motbob) [reply lvsfunc/comparison.py:612]: thr was designed to be able to be negative. so this is fixing a bug,

File: lvsfunc/comparison.py (hunk 2 above)
Comment (Setsugennoao) [reply lvsfunc/comparison.py:613]: You should specify why it errored, aka passing `reason=thr`.

File: lvsfunc/comparison.py (hunk 4 above)
Comment (motbob) [reply lvsfunc/comparison.py:616]: sorry, I don't understand. note that i explained below that these values were derived experimentally.

File: lvsfunc/comparison.py (hunk 3 above)
Comment (Setsugennoao) [reply lvsfunc/comparison.py:612]: Then check if abs(thr) <= 128 or ths is in [-128, 128]

Comment (motbob) [reply lvsfunc/comparison.py:612]: Oh, I'm sorry, I got it wrong. It should check for 0 <= thr < 128 (or perhaps 0 < thr < 128). I will check and fix it.

Comment (Setsugennoao) [reply lvsfunc/comparison.py:612]: Gotcha, thanks.
You can resolve the convo after committing.

**Comment by LightArrowsEXE** [comment]
Is this still getting looked at?

**Comment by Setsugennoao** [comment]
I guess I'll have to take over this PR, since op is busy.
